online-shop/src/app/components/containers/product-details/product-details.component.ts

```diff
@@ -1,23 +1,33 @@
-import { Component, Input } from '@angular/core';
+import { Component, OnInit } from '@angular/core';
 import { ActivatedRoute } from '@angular/router';
-import { products } from 'src/app/mocks/products.mocks';
 import { Product } from 'src/app/modules/shared/types/products.types';
+import { ProductService } from 'src/app/services/products.service';
+
 @Component({
   selector: 'app-product-details',
   templateUrl: './product-details.component.html',
   styleUrls: ['./product-details.component.scss'],
 })
-export class ProductsDetailsComponent {
-  product!:Product
+export class ProductDetailsComponent implements OnInit {
+  product!: Product;
+
   constructor(
-    private route:ActivatedRoute
-  ){}
-  ngOnInit():void{
-    this.getProduct();
-  }
-  getProduct():void{
-    const id=this.route.snapshot.paramMap.get('id');
-    this.product = products.find(p => p.id === id) ?? {} as Product;
+    private route: ActivatedRoute,
+    private productService: ProductService
+  ) {}
 
+  ngOnInit() {
+    this.route.params.subscribe(params => {
+      const productId = params['id'];
+      this.productService.getProduct(productId).subscribe(product => {
```

Comment:
Make sure you unsubscribe from each subscription you made

Reply:
I now use @UntilDestroy({ checkProperties: true })